File: samples/Microsoft.AspNet.SignalR.Hosting.AspNet.Samples/Hubs/HubClientsAPIs/HubClientsAPIs.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNet.SignalR.Hubs;

namespace Microsoft.AspNet.SignalR.Hosting.AspNet.Samples.Hubs.HubClientsAPIs
{
    public class HubClientsAPIs : Hub
    {
        public void DisplayMessageAll(string connectionId, string message)
        {
            Clients.All.displayMessage("From Clients.All: " + message + " " + connectionId);
        }

        public void DisplayMessageAllExcept(string connectionId, string message, params string[] target_connectionId)
        {
            Clients.AllExcept(target_connectionId).displayMessage("From Clients.AllExcept: " + message + " " + connectionId);
        }

        public void DisplayMessageOther(string connectionId, string message)
        {
            Clients.Others.displayMessage("From Clients.Others: " + message + " " + connectionId);
        }

        public void DisplayMessageCaller(string connectionId, string message)
        {
            Clients.Caller.displayMessage("From Clients.Caller: " + message + " " + connectionId);
        }

        public void DisplayMessageSpecified(string connectionId, string target_connectionId, string message)
        {
            Clients.Client(target_connectionId).displayMessage("From Clients.Client: " + message + " " + connectionId);
        }

        public string JoinGroup(string connectionId, string groupNme)
        {
            Groups.Add(connectionId, groupNme);
            return connectionId + " joined " + groupNme;
        }

        public string LeaveGroup(string connectionId, string groupNme)
        {
            Groups.Remove(connectionId, groupNme);
            return connectionId + " removed " + groupNme;
        }

        public void DisplayMessageGroup(string connectionId, string groupNme, string message)
        {
            Clients.Group(groupNme, "").displayMessage("From Clients.Group: " + message + " " + connectionId);
        }
        
        public void DisplayMessageOthersInGroup(string connectionId, string groupNme, string message)
        {
            Clients.OthersInGroup(groupNme).displayMessage("From Clients.OthersInGroup: " + message + " " + connectionId);
        }
        
        public override Task OnConnected()
        {
            Clients.All.displayMessage(Context.ConnectionId + " OnConnected");
            return null;
        }

        public override Task OnReconnected()
        {
            Clients.Caller.displayMessage(Context.ConnectionId + " OnReconnected");
            return null;
        }

        public override Task OnDisconnected()
        {
            Clients.All.displayMessage(Context.ConnectionId + " OnDisconnected");
            return null;
        }
    }
}

File: samples/Microsoft.AspNet.SignalR.Hosting.AspNet.Samples/Hubs/HubClientsAPIs/HubClientsAPIs.js
/// <reference path="../../Scripts/jquery-1.8.2.js" />
$(function () {
    "use strict";

    var clientsAPIsHub = $.connection.hubClientsAPIs,
        messages = $("#messages"),
        stopStartBtn = $("#stopStart"),
        start;
    $.connection.hub.logging = true;

    clientsAPIsHub.client.displayMessage = function (value) {
        $("<li/>").html("[" + new Date().toTimeString() + "]: " + value).appendTo(messages);
    }

    $.connection.hub.stateChanged(function (change) {
        var oldState = null,
            newState = null;
        for (var p in $.signalR.connectionState) {
            if ($.signalR.connectionState[p] === change.oldState) {
                oldState = p;
            } else if ($.signalR.connectionState[p] === change.newState) {
                newState = p;
            }
        }

        $("<li/>").html("[" + new Date().toTimeString() + "]: " + oldState + " => " + newState + " " + $.connection.hub.id)
                    .appendTo(messages);
    });

    $.connection.hub.reconnected(function () {
        $("<li/>").css("background-color", "green")
                    .css("color", "white")
                    .html("[" + new Date().toTimeString() + "]: Connection re-established")
                    .appendTo(messages);
        $("<li/>").html("reconnected transport: " + $.connection.hub.transport.name + " " + $.connection.hub.id)
                    .appendTo(messages);
    });

    $.connection.hub.error(function (err) {
        $("<li/>").html("Error occurred: " + err).appendTo(messages);
    });

    $.connection.hub.connectionSlow(function () {
        $("<li/>").html("[" + new Date().toTimeString() + "]: Connection Slow").appendTo(messages);
    });

    $.connection.hub.disconnected(function () {
        stopStartBtn.prop("disabled", false)
                    .find("span")
                        .text("Start Connection")
                        .end()
                    .find("i")
                        .removeClass("icon-stop")
                        .addClass("icon-play");
    });

    start = function () {
        $.connection.hub.start({ transport: activeTransport, jsonp: isJsonp })
            .done(function () {
                $("<li/>").html("started transport: " + $.connection.hub.transport.name + " " + $.connection.hub.id)
                            .appendTo(messages);

                stopStartBtn.prop("disabled", false)
                            .find("span")
                                .text("Stop Connection")
                                .end()
                            .find("i")
                                .removeClass("icon-play")
                                .addClass("icon-stop");

            });
    };
    start();


    $('#join-group').click(function () {
        var connectionIdToJoin = $.connection.hub.id;

        if ($("#connection").val())
            connectionIdToJoin = $("#connection").val();

        clientsAPIsHub.server.joinGroup(connectionIdToJoin, $("#group").val()).done(function (value1) {
            $("<li/>").html("Succeeded at joinGroup: " + value1).appendTo(messages);
        }).fail(function (e) {
            $("<li/>").html("Failed at joinGroup: " + e).appendTo(messages);
        });
    });

    $('#leave-group').click(function () {
        var connectionIdToLeave = $.connection.hub.id;

        if ($("#connection").val())
            connectionIdToLeave = $("#connection").val();

        clientsAPIsHub.server.leaveGroup(connectionIdToLeave, $("#group").val()).done(function (value1) {
            $("<li/>").html("Succeeded at leaveGroup: " + value1).appendTo(messages);
        }).fail(function (e) {
            $("<li/>").html("Failed at leaveGroup: " + e).appendTo(messages);
        });
    });

    $("#broadcast").click(function () {
        clientsAPIsHub.server.displayMessageAll($.connection.hub.id, $("#msg").val()).fail(function (e) {
            $("<li/>").html("Failed at getMessage: " + e).appendTo(messages);
        });
    });

    $("#broadcast-except-specified").click(function () {
        clientsAPIsHub.server.displayMessageAllExcept($.connection.hub.id, $("#msg").val(), $("#connection").val().split(",")).fail(function (e) {
            $("<li/>").html("Failed at getMessageAllExcept: " + e).appendTo(messages);
        });
    });


    $("#other").click(function () {
        clientsAPIsHub.server.displayMessageOther($.connection.hub.id, $("#msg").val()).fail(function (e) {
            $("<li/>").html("Failed at getMessageOther: " + e).appendTo(messages);
        });
    });


    $("#sendToMe").click(function () {
        clientsAPIsHub.server.displayMessageCaller($.connection.hub.id, $("#me").val()).fail(function (e) {
            $("<li/>").html("Failed at getMessageCaller: " + e).appendTo(messages);
        });
    });


    $("#specified").click(function () {
        clientsAPIsHub.server.displayMessageSpecified($.connection.hub.id, $("#connection").val(), $("#me").val()).fail(function (e) {
            $("<li/>").html("Failed at getMessageSpecified: " + e).appendTo(messages);
        });
    });

    $("#groupmsg").click(function () {
        clientsAPIsHub.server.displayMessageGroup($.connection.hub.id, $("#group").val(), $("#groupMessage").val()).fail(function (e) {
            $("<li/>").html("Failed at getMessageGroup: " + e).appendTo(messages);
        });
    });

    $("#otherInGroupmsg").click(function () {
        clientsAPIsHub.server.displayMessageOthersInGroup($.connection.hub.id, $("#group").val(), $("#groupMessage").val()).fail(function (e) {
            $("<li/>").html("Failed at getMessageOthersInGroup: " + e).appendTo(messages);
        });
    });

    stopStartBtn.click(function () {
        var $el = $(this);

        $el.prop("disabled", true);

        if ($.trim($el.find("span").text()) === "Stop Connection") {
            $.connection.hub.stop();
        } else {
            start();
        }
    });
});
